src/pudl_archiver/zenodo/api_client.py

```diff
@@ -196,32 +211,43 @@ async def add_files(self, resources: dict[str, ResourceInfo]):
             Updated Deposition object.
         """
         for name, resource in resources.items():
+            filepath = resource.local_path.name
             if name not in self.deposition_files:
-                self.changed = True
                 logger.info(f"Adding {name} to deposition.")
 
-                with open(resource.local_path, "rb") as f:
-                    await self.upload(f, resource.local_path.name)
+                self.changeset["upload"].append((Path(filepath), filepath))
             else:
                 file_info = self.deposition_files[name]
 
                 # If file is not exact match for existing file, update with new file
                 if not _compute_md5(resource.local_path) == file_info.checksum:
-                    self.changed = True
                     logger.info(f"Updating {name}")
-                    await self.delete_file(file_info)
-
-                    with open(resource.local_path, "rb") as f:
-                        await self.upload(f, resource.local_path.name)
+                    self.changeset["delete"].append(file_info)
+                    self.changeset["upload"].append((Path(filepath), filepath))
 
         # Delete files not included in new deposition
-        for filename, file in self.deposition_files.items():
+        for filename, file_info in self.deposition_files.items():
             if filename not in resources and filename != "datapackage.json":
-                self.changed = True
-                await self.delete_file(file)
+                self.changeset["delete"].append(file_info)
 
         await self.update_datapackage(resources)
 
+        await self._apply_changeset()
+
+    async def _apply_changeset(self):
```

Comment:
Probably not too big of a deal since this isn't in the public API, but a one sentence docstring could be nice